prov/util/src/util_mem_monitor.c

```diff
@@ -86,3 +89,143 @@ void ofi_monitor_unsubscribe(struct ofi_mem_monitor *monitor,
 	       "unsubscribing addr=%p len=%zu\n", addr, len);
 	monitor->unsubscribe(monitor, addr, len);
 }
+
+#if HAVE_UFFD_UNMAP
+
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/userfaultfd.h>
+
+
+static int ofi_uffd_register(const void *addr, size_t len, size_t page_size)
+{
+	struct uffdio_register reg;
+	int ret;
+
+	reg.range.start = (uint64_t) (uintptr_t)
+			  ofi_get_page_start(addr, page_size);
+	reg.range.len = ofi_get_page_bytes(addr, len, page_size);
+	reg.mode = UFFDIO_REGISTER_MODE_MISSING;
+	ret = ioctl(uffd.fd, UFFDIO_REGISTER, &reg);
+	if (ret < 0) {
+		if (errno != EINVAL) {
+			FI_WARN(&core_prov, FI_LOG_MR,
+				"ioctl/uffd_unreg: %s\n", strerror(errno));
+		}
+		return -errno;
+	}
+	return 0;
+}
+
+static int ofi_uffd_subscribe(struct ofi_mem_monitor *monitor,
+			      const void *addr, size_t len)
+{
+	int ret;
+
+	assert(monitor == &uffd.monitor);
+	ret = ofi_uffd_register(addr, len, uffd.page_size);
+	if (ret)
+		ret = ofi_uffd_register(addr, len, uffd.hugepage_size);
+	return ret;
+}
+
+static int ofi_uffd_unregister(const void *addr, size_t len, size_t page_size)
+{
+	struct uffdio_range range;
+	int ret;
+
+	range.start = (uint64_t) (uintptr_t)
+		      ofi_get_page_start(addr, page_size);
+	range.len = ofi_get_page_bytes(addr, len, page_size);
+	ret = ioctl(uffd.fd, UFFDIO_UNREGISTER, &range);
+	if (ret < 0) {
+		if (errno != EINVAL) {
+			FI_WARN(&core_prov, FI_LOG_MR,
+				"ioctl/uffd_unreg: %s\n", strerror(errno));
+		}
+		return -errno;
+	}
+	return 0;
+}
+
+static void ofi_uffd_unsubscribe(struct ofi_mem_monitor *monitor,
+				 const void *addr, size_t len)
+{
+	int ret;
+
+	assert(monitor == &uffd.monitor);
+	ret = ofi_uffd_unregister(addr, len, uffd.page_size);
+	if (ret)
+		ret = ofi_uffd_unregister(addr, len, uffd.hugepage_size);
```

Comment:
Is there a reason why we try regular page_size first?

Reply:
It matches the subscribe call, but the checks are optimized assuming regular page sizes first.  Maybe this could be more intelligent and try huge pages first if the len is >= hugepage_size?

Reply:
I'm not completely familiar with userfaultfd but assuming registering for hugepage_size (after checking length) gives better performance in terms of the kernel monitoring and notification, I suppose that would be preferable.

Reply:
What this is doing is registering the memory buffer provided by the app with uffd.  The registration must be along a page boundary.  What we don't know is if the buffer is backed by normal pages or huge pages.  The code above first checks if the buffer is backed by normal pages.  If that fails, it tries huge pages.  So, we're not selecting which one this is.  We're only trying to guess what type of pages back the buffer we were given, and we're expecting the kernel to return a failure if we guess wrong.

What we need here is a way to determine what type of pages are behind the virtual address.  But I don't know how to get that.